---
title: "Wang et al Comparison"
output: 'github_document'
---

* See [Mixed infections in genotypic drug-resistant Mycobacterium tuberculosis](https://doi.org/10.1038/wang-023-44341-x)

```{r setup}
library(tidyverse)
ref_fn <- '01_data/GCA_000195955.2_ASM19595v2_genomic.fna.gz'

if (!dir.exists('06_data')) { dir.create('06_data') }
```

```{r save_wang_supp_1, eval=FALSE}
# from https://www.nature.com/articles/s41598-023-44341-x
read_csv('https://static-content.springer.com/esm/art%3A10.1038%2Fs41598-023-44341-x/MediaObjects/41598_2023_44341_MOESM1_ESM.csv') %>% 
  left_join(
    # Note: List of all strains detected as mixed by TBProfiler, provided by author's in private communication
    read_csv('06_data/MSI_accession.csv', col_names = 'wgs_id') %>% 
      select(wgs_id) %>% 
      mutate(is_mixed = TRUE)
  ) %>% 
  mutate(is_mixed = replace_na(is_mixed, FALSE)) %>%
  saveRDS('06_data/wang_supp_1.rds')
```

```{r get_subset, eval=FALSE}
# wang_supp_1 <- readRDS('06_data/wang_supp_1.rds')
# only run once to select 5000 samples at random
# wang_supp_1_n5000 <- sample_n(wang_supp_1, 5000)
# saveRDS(wang_supp_1_n5000, '06_data/wang_supp_1_n5000.rds')

wang_supp_1_n5000 <- readRDS('06_data/wang_supp_1_n5000.rds')

wang_supp_1_n5000 %>%
  pull(wgs_id) %>%
  write_lines('06_data/fetchngs/ids.csv')
```

```{bash run_fetch_ngs, eval=FALSE}
cd 06_data/fetchngs/
nextflow run nf-core/fetchngs --input ids.csv --outdir output --download_method aspera
```

```{r setup_tbtyper, eval=FALSE}
wang_supp_1_n5000 <- readRDS('06_data/wang_supp_1_n5000.rds')

ss <- 
  read_csv('06_data/fetchngs/output/samplesheet/samplesheet.csv') %>% 
  bind_rows(read_csv('06_data/fetchngs/output2/samplesheet/samplesheet.csv'))

manifest <-
  bind_rows(
    wang_supp_1_n5000 %>% 
      mutate(run_accession = wgs_id) %>% 
      inner_join(ss),
    wang_supp_1_n5000 %>% 
      mutate(sample_accession = wgs_id) %>% 
      inner_join(ss) %>% 
      group_by(sample_accession) %>% 
      slice(which.max(base_count))
  ) %>% 
  select(wgs_id, fastq_1, fastq_2) %>% 
  mutate(across(starts_with('fastq'), ~ normalizePath(str_c('./06_data/fetchngs/',.),  mustWork = F))) %>% 
  mutate(fastq_2 = if_else(basename(fastq_2) == 'NA', NA_character_, fastq_2)) %>%
  filter(file.exists(fastq_1))

if (!dir.exists('06_data/tbtyper')) {
  dir.create('06_data/tbtyper')
}
# copy nextflow config
read_lines('benchmark_default.config')  %>% 
  str_replace('params.methods.*', 'params.methods = [\'FastTBtypeR\']\n') %>% 
  write_lines('06_data/tbtyper/nextflow.config')

# set params.max_mix = 2

# copy ref
file.copy(ref_fn, '06_data/tbtyper/')

manifest %>% 
  select(sample = wgs_id, fastq1 = fastq_1, fastq2 = fastq_2) %>% 
  write_tsv('06_data/tbtyper/manifest.tsv')
```

```{bash run_tbtyper, eval=FALSE}
cd 06_data/tbtyper/
nextflow run ../../../TBtypeNF/main.nf
```

```{r save_results, eval=FALSE}
res_wang_subset_raw <- read_tsv('06_data/tbtyper/output/combined_results.tsv')
saveRDS(res_wang_subset_raw, '06_data/res_wang_subset_raw.rds')
# use as Table S2
```


```{r}
res_wang_subset_raw <- readRDS('06_data/res_wang_subset_raw.rds')

res_wang_subset <-
  res_wang_subset_raw %>% 
  mutate(proportion = replace_na(proportion, 1)) %>% 
  arrange(sample, desc(proportion)) %>% 
  filter(proportion >= 0.01) %>%
  group_by(sample) %>% 
  summarise(
    minor_prop = sum(proportion[proportion!=max(proportion)]),
    strains = str_c(strain, ' [', round(100*proportion, 1), '%]', collapse = '\n')) %>% 
  mutate(is_mixed = minor_prop >0) %>% 
  mutate(is_mixed = replace_na(is_mixed))

wang_panel_A_dat <- 
  res_wang_subset %>% 
  select(sample, minor_prop, is_mixed) %>% 
  inner_join(
    wang_supp_1_n5000 %>% 
      select(sample = wgs_id, is_mixed),
    by = 'sample',
    suffix = c('_TBtypeR', '_TBProfiler')
  ) %>% 
  mutate(set = case_when(
    is_mixed_TBtypeR & is_mixed_TBProfiler ~ 'Both',
    is_mixed_TBtypeR                       ~ 'TBtypeR only',
    is_mixed_TBProfiler                    ~ 'TBProfiler only',
  )) %>% 
  mutate(MSF = 100 * minor_prop) %>% 
  filter(!is.na(set)) %>% 
  rename(Tool = set) %>% 
  mutate(Tool = factor(Tool, c('TBtypeR only', 'Both')))

saveRDS(wang_panel_A_dat, '06_data/wang_panel_A_dat.rds')

# use in 09_Figure_2.Rmd
```


```{r stats}
# binomial conf int.
res_wang_subset %>% 
  count(is_mixed) %>% 
  mutate(p = 100 * (n / sum(n))) %>% 
  filter(is_mixed) %>% 
  rowwise() %>% 
  mutate(
    lower = 100 * binom.test(n, 5000)$conf.int[1],
    upper = 100 * binom.test(n, 5000)$conf.int[2])

# conunt below MSF % per tool
wang_panel_A_dat %>%
  group_by(Tool) %>%
  count(MSF < 5) %>% 
  mutate(p = 100 * (n / sum(n)))
```

```{r}
read_csv('06_data/fetchngs/baf_samples/samplesheet/samplesheet.csv') %>% 
  select(sample = run_accession, fastq1=fastq_1, fastq2=fastq_2) %>% 
  mutate(across(starts_with('fastq'), ~ normalizePath(str_c('./06_data/fetchngs/', .)))) %>% 
  write_tsv('06_data/tbtyper/baf_samples_manifest.tsv')
```

```{r extract_BAF_subset, eval=FALSE}
baf_ids <-
  res_wang_subset %>%
  select(sample, minor_prop, is_mixed, strains) %>%
  inner_join(
    wang_supp_1_n5000 %>%
      mutate(is_mixed = str_detect(main_lineage, ';')) %>%
      select(sample = wgs_id, is_mixed),
    by = 'sample',
    suffix = c('_TBtypeR', '_TBProfiler')
  ) %>%
  filter(!is.na(is_mixed_TBProfiler)) %>%
  mutate(set = case_when(
    is_mixed_TBtypeR & is_mixed_TBProfiler ~ 'Both',
    is_mixed_TBtypeR                       ~ 'TBtypeR',
    is_mixed_TBProfiler                    ~ 'TBProfiler',
  )) %>%
  mutate(freq = if_else(minor_prop < 0.05, 'below 5%', 'above 5%')) %>%
  mutate(set =
           str_c(set, '\n', freq) %>%
           factor(c('Both\nabove 5%', 'TBtypeR\nabove 5%', 'TBtypeR\nbelow 5%'))) %>%
  filter(!is.na(set)) %>%
  group_by(set, freq) %>%
  arrange(minor_prop) %>%
  mutate(bin = 1 + (row_number()-1) %/% (n()/5)) %>%
  arrange(sample) %>%
  group_by(set, freq, bin) %>%
  (\(x) { set.seed(42); sample_n(x, 1) } ) %>%
  ungroup()

trace <- read_tsv('06_data/tbtyper/log.txt', col_names = c('name', 'dir'))

sm_baf <-
  trace %>%
  filter(str_detect(name, 'FASTLIN')) %>%
  filter(str_detect(name, str_c(baf_ids$sample, collapse = '|'))) %>%
  mutate(sample = str_extract(name, '(?<=\\().+(?=\\))'),
         file = str_c(dir, '/', sample, '_fastlin.tsv.gz')) %>%
  select(sample, file) %>%
  mutate(ac = map2(sample, file, \(s, f) {
    ac <- TBtypeR:::fastlin_allele_counts(f, sample = s, panel = TBtypeR::tbt_panel)
    ref <- ac[,,'ref']
    alt <- ac[,,'alt']
    tibble(sample = s,
           variant = names(ref),
           ref = unname(ref),
           alt = unname(alt),
           baf = alt / (ref + alt))
  })) %>%
  select(ac) %>%
  unnest(ac)

sm_baf_mod <-
  res_wang_subset_raw %>%
  select(sample, strain, proportion) %>%
  semi_join(sm_baf, by = 'sample') %>%
  inner_join(
    TBtypeR:::panel_to_geno(panel = TBtypeR::tbt_panel) %>%
      as.data.frame() %>%
      rownames_to_column('strain') %>%
      as_tibble() %>%
      pivot_longer(-strain, names_to = 'variant', values_to = 'genotype'),
    by = c('strain'),
    relationship = 'many-to-many'
  ) %>%
  semi_join(sm_baf, by = c('sample', 'variant')) %>%
  arrange(sample, variant, desc(proportion)) %>%
  group_by(sample, variant) %>%
  summarise(
    state = str_c(genotype, collapse = ''),
    exp_baf = sum(proportion * genotype),
    .groups = 'drop'
  ) %>%
  inner_join(sm_baf, by = c('sample', 'variant'))

wang_panel_B_dat <-
  sm_baf_mod %>%
  left_join(
     select(baf_ids, sample, set, strains, bin),
     by = 'sample') %>%
  mutate(sample = str_c(sample, '\n', strains)) %>%
  arrange(desc(set), bin, sample) %>%
  mutate(sample = as_factor(sample),
         state = as.factor(state)) %>%
  select(sample, set, state, exp_baf, baf)

saveRDS(wang_panel_B_dat, '06_data/wang_panel_B_dat.rds')

# use in 09_Figure_2.Rmd
```




